### PR TITLE
Models

### PR DESCRIPTION
**Pull request by sofiaele**
This branch contains models folder for train and test text.
You need to download pretrained fasttext model in wikipedia data from here: [https://fasttext.cc/docs/en/pretrained-vectors.html](url) (English bin+text) , extract them and use the .vec file. 

For training: 

- from train_text run fast_text_and_svm function with input (your data, .vec pretrained model). The model will be saved with name "svm_text_model"  and additionally a csv file with the names of the classes will be saved ("svm_text_classes.csv").
P.S. You can find "your data" input here [https://drive.google.com/drive/folders/1FmesRxmxooICePJ-3gPQ6E9g7EQgkMpE?usp=sharing](url)  (transcriptions + labels)

For testing: 

- First run training to make model or use a model already trained from here: [https://drive.google.com/drive/folders/1FmesRxmxooICePJ-3gPQ6E9g7EQgkMpE?usp=sharing](url) 
- from test_text run predict_text_labels with input (your data, .vec pretrained model, svm model, csv file of classes). A dictionary will be returned which has as elements the classes and as values the percentage that your data is classified (belongs) to each class.

### CODE REVIEW

**Comment by sofiaele** [comment]
For training text: 
1. Download fasttext pretrained model : wget https://dl.fbaipublicfiles.com/fasttext/vectors-wiki/wiki.en.zip 
2. Unzip downloaded file : unzip wiki.en.zip 
3. Run train text : python3 train_text.py -a "input" -p "wiki.en.bin" -o "svm_output_classifier_path" 
Where **input** is input data in csv format with one column "transcriptions" and one column "labels" . You can find such csv files here: [https://drive.google.com/drive/folders/1FmesRxmxooICePJ-3gPQ6E9g7EQgkMpE?usp=sharing](url).
The output classifier will be saved at the defined "svm_output_classifier_path".

For testing text:
1.Run test_text : python3 test_text.py -i "input_string" -p "wiki.en.bin" -c "svm_classifier_path" -n "classes_names" 
Where "input_string" is the input text that we want to classify , "wiki.en.bin" is the path of the pretrained fasttext model , "svm_classifier_path" is the path of the classifier we trained, "classes_names" is a csv file with the names of the classes(saved automatically when trained). 

For training audio: 
1. python3 train_audio.py -i "input_path" -o "svm_output_classifier_path" 
Where "input_path" is the path of the directory which contains audio organized in folders of classes --you can find such organized input audio here [https://drive.google.com/drive/folders/12Ai25ZUCysAUUWRSiSd5WrsWRbB1gRuF?usp=sharing](url)-- , "svm_output_classifier_path" the path to which model will be saved. 

For testing audio: 
1. python3 test_audio.py -i "input_wav" -c "svm_classifier_path"

**Comment by lobracost** [comment]
@sofiaele  @tyiannak Everything seems to work as expected. However, I have to propose some future improvements.

### General comments

1. The matrix data structure is mostly used on previous versions of Python. For that reason I suggest to change the commands np.asmatrix to np.asarray and np.matrix to np.array.
2. Maybe use fastText from the TorchText framework. This way it would be easier to run the models using cuda.


### train_text

Works fine.

Proposed changes:

1. Lemmatization of words would maybe help in the text_preprocess function.
2. The normalization of features is too complicated. I propose to replace these lines https://github.com/tyiannak/readys/blob/7d7b8fd8103860959a0f5662e7b7da90afe22494/models/train_text.py#L45-L51 with `normalized_features = np.array([(vec - mean)/std for vec in X])`
3. These lines https://github.com/tyiannak/readys/blob/7d7b8fd8103860959a0f5662e7b7da90afe22494/models/train_text.py#L132-L146 is better to lie under a separate function which loads the dataset.
4. We must consider the case of a sentence for which the fasttext model cannot extract features. This case can occur in a range of different scenarios. In such situations, the current version of the code will throw an error.

### test_text

Works fine.

Proposed changes: None


### train_audio

Works fine.

Proposed changes:

1. Include a config file which contains the input values for the pyaudio functions
2. Write our own train function and don't use the pyaudio one. This way we can modify it as we wish.


### test_audio

Works fine.

Proposed changes: None
